network/manager.go

```diff
@@ -0,0 +1,296 @@
+	}
+
+	var eaddr net.IP
+
+	if len(opts.publicIP) > 0 {
+		eaddr = net.ParseIP(opts.publicIP)
+		if eaddr == nil {
+			return nil, fmt.Errorf("Invalid public IP address", opts.publicIP)
+		}
+	}
+
+	if eaddr == nil {
+		eaddr = iaddr
+	}
+
+	log.Infof("Using %s as external interface", iaddr)
+	log.Infof("Using %s as external endpoint", eaddr)
+
+	manager := &Manager{
+		sm:              sm,
+		allowedNetworks: make(map[string]bool),
+		networks:        make(map[string]*Network),
+		watch:           opts.watchNetworks,
+		ipMasq:          opts.ipMasq,
+		extIface:        iface,
+		iaddr:           iaddr,
+		eaddr:           eaddr,
+	}
+
+	for _, name := range strings.Split(opts.networks, ",") {
+		manager.allowedNetworks[name] = true
+	}
+
+	if manager.isMultiNetwork() {
+		// Get list of existing networks
+		result, err := manager.sm.WatchNetworks(ctx, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, n := range result.Snapshot {
+			if manager.isNetAllowed(n) {
+				manager.networks[n] = NewNetwork(sm, n, manager.ipMasq)
+			}
+		}
+	} else {
+		manager.networks[""] = NewNetwork(sm, "", manager.ipMasq)
+	}
+
+	return manager, nil
+}
+
+func lookupExtIface(ifname string) (*net.Interface, net.IP, error) {
+	var iface *net.Interface
+	var iaddr net.IP
+	var err error
+
+	if len(ifname) > 0 {
+		if iaddr = net.ParseIP(ifname); iaddr != nil {
+			iface, err = ip.GetInterfaceByIP(iaddr)
+			if err != nil {
+				return nil, nil, fmt.Errorf("Error looking up interface %s: %s", ifname, err)
+			}
+		} else {
+			iface, err = net.InterfaceByName(ifname)
+			if err != nil {
+				return nil, nil, fmt.Errorf("Error looking up interface %s: %s", ifname, err)
+			}
+		}
+	} else {
+		log.Info("Determining IP address of default interface")
+		if iface, err = ip.GetDefaultGatewayIface(); err != nil {
+			return nil, nil, fmt.Errorf("Failed to get default interface: %s", err)
+		}
+	}
+
+	if iaddr == nil {
+		iaddr, err = ip.GetIfaceIP4Addr(iface)
+		if err != nil {
+			return nil, nil, fmt.Errorf("Failed to find IPv4 address for interface %s", iface.Name)
+		}
+	}
+
+	return iface, iaddr, nil
+}
+
+func writeSubnetFile(path string, nw ip.IP4Net, ipMasq bool, sn *backend.SubnetDef) error {
+	dir, name := filepath.Split(path)
+	os.MkdirAll(dir, 0755)
+
+	tempFile := filepath.Join(dir, "."+name)
+	f, err := os.Create(tempFile)
+	if err != nil {
+		return err
+	}
+
+	// Write out the first usable IP by incrementing
+	// sn.IP by one
+	sn.Net.IP += 1
+
+	fmt.Fprintf(f, "FLANNEL_NETWORK=%s\n", nw)
+	fmt.Fprintf(f, "FLANNEL_SUBNET=%s\n", sn.Net)
+	fmt.Fprintf(f, "FLANNEL_MTU=%d\n", sn.MTU)
+	_, err = fmt.Fprintf(f, "FLANNEL_IPMASQ=%v\n", ipMasq)
+	f.Close()
+	if err != nil {
+		return err
+	}
+
+	// rename(2) the temporary file to the desired location so that it becomes
+	// atomically visible with the contents
+	return os.Rename(tempFile, path)
+}
+
+func (m *Manager) RunNetwork(net *Network, wg sync.WaitGroup) {
+	wg.Add(1)
+	defer wg.Done()
```

Comment:
Stylistic nit: so far I have been following the convention of putting the wait group's add/done calls outside:

```
go func() {
    wg.Add(1)
    defer wg.Done()
    nm.RunNetwork(net)
}
```

I'm not sure if it's "the right way" but would like to keep it for consistency.